query_optimizer/logical/HashJoin.hpp

```diff
@@ -132,16 +181,26 @@ class HashJoin : public BinaryJoin {
   HashJoin(const LogicalPtr &left,
            const LogicalPtr &right,
            const std::vector<expressions::AttributeReferencePtr> &left_join_attributes,
-           const std::vector<expressions::AttributeReferencePtr> &right_join_attributes)
+           const std::vector<expressions::AttributeReferencePtr> &right_join_attributes,
+           const expressions::PredicatePtr &residual_predicate,
+           const JoinType join_type)
       : BinaryJoin(left, right),
         left_join_attributes_(left_join_attributes),
-        right_join_attributes_(right_join_attributes) {
+        right_join_attributes_(right_join_attributes),
+        residual_predicate_(residual_predicate),
+        join_type_(join_type) {
     DCHECK_EQ(left_join_attributes.size(), right_join_attributes.size());
     DCHECK(!left_join_attributes.empty());
+
+    if (residual_predicate_ != nullptr) {
+      addInputExpression(residual_predicate_);
+    }
   }
 
   std::vector<expressions::AttributeReferencePtr> left_join_attributes_;
   std::vector<expressions::AttributeReferencePtr> right_join_attributes_;
+  expressions::PredicatePtr residual_predicate_;
```

Comment:
In fact all member variables of `Logical` subclasses can be `const` because the only way to access a `Logical` node is via `std::shard_ptr<const SubclassOfLogical>`. So in terms of readability, we can just avoid the redundancy by decorating none of them. Or we may declare all member variables of all subclasses of `Logical` to be `const`.